### PR TITLE
CONTRIBUTING: Update/Modernize

### PR DESCRIPTION
**Pull request by ax3l**
- Add GitHub account setup
- Add local git setup
- Modernize for CMake workflows

Same as https://github.com/ECP-WarpX/impactx/pull/16

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
-on the `WarpX Github page <https://github.com/ECP-WarpX/WarpX>`_, press the fork button.
-Then, you can execute:
+   - ``git config --global user.name "FIRSTNAME LASTNAME"``
+   - ``git config --global user.email EMAIL@EXAMPLE.com``
+4. Go to https://github.com/settings/keys and add the SSH public key of the machine you develop on. (Check out the GitHub guide to `generating SSH keys <https://docs.github.com/articles/generating-an-ssh-key/>`__ or `troubleshoot common SSH problems <https://docs.github.com/ssh-issues/>`__. )
+
+Make your own fork
+^^^^^^^^^^^^^^^^^^
+
+First, fork the WarpX `"mainline" repo on GitHub <https://github.com/ECP-WarpX/WarpX>`__ by pressing press the *Fork* button on the top right of the page.
```
Comment (EZoni) [inline CONTRIBUTING.rst]:
```suggestion
First, fork the WarpX `"mainline" repo on GitHub <https://github.com/ECP-WarpX/WarpX>`__ by pressing the *Fork* button on the top right of the page.
```

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
+Then, we create local copies, for development:
 
 .. code-block:: sh
 
-   # These 4 first lines are the same as for a standard WarpX install
-   mkdir warpx_directory
-   cd warpx_directory
-   git clone --branch development https://github.com/ECP-WarpX/picsar.git
-   git clone --branch development https://github.com/AMReX-Codes/amrex.git
+   # Clone your the mainline WarpX source code to your local computer.
```
Comment (EZoni) [inline CONTRIBUTING.rst]:
```suggestion
   # Clone the mainline WarpX source code to your local computer.
```

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
-   You do not have to re-do the setup above every time.
-   Instead, in the future, all you need is to update the ``development`` branch on your fork with:
+   We only need to do the above steps for the first time.
+
+Let's Develop
+^^^^^^^^^^^^^
 
-   .. code-block:: sh
+You are all set!
+Now, he basic WarpX development workflow is:
```
Comment (EZoni) [inline CONTRIBUTING.rst]:
```suggestion
Now, the basic WarpX development workflow is:
```

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
 
 .. code-block:: sh
 
+   # start from an up-to-date development branch
    git checkout development
+   git pull mainline development
 
-in the WarpX directory.
+   # create a fresh branch
+   git checkout -b <branch_name>
```
Comment (EZoni) [inline CONTRIBUTING.rst]: Is it possible that someone new to GitHub misinterpret these `<` and `>` as mandatory in the Git syntax? Maybe not, but I'm wondering if it's easier to simply write `git checkout -b branch_name`, without the `<` and `>`.

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
 
-   git checkout -b <branch_name>
+* WarpX Doxygen: https://warpx.readthedocs.io/en/latest/_static/doxyhtml
+* AMReX Doxygen: https://amrex-codes.github.io/amrex/doxygen
+* PICSAR Doxygen: (todo)
 
-and do the coding you want.
-It is probably a good time to look at the `AMReX documentation <https://amrex-codes.github.io/amrex/docs_html/>`_ and at the `AMReX Doxygen reference <https://ccse.lbl.gov/pub/AMReX_Docs/index.html>`_.
-Add the files you work on to the git staging area with
+Once you are done developing, add the files you modified to the ``git`` *staging area* with
```
Comment (EZoni) [inline CONTRIBUTING.rst]:
```suggestion
Once you are done developing, add the files you created and/or modified to the ``git`` *staging area* with
```

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
+
+   This is a short, 40-character title
+
+   After a newline, you can write arbitray paragraphs. You
+   usually limit the lines to 70 characters, but if you don't, then
+   nothing bad will happen.
 
-The commit message (between quotation marks) is super important in order to follow the developments and identify bugs.
+   The most important part is really that you find a descriptive title
+   and add an empty newline after it.
```
Comment (EZoni) [inline CONTRIBUTING.rst:126]: Here we could add something like: "If you realize that you made a mistake in your commit message and want to change it, before pushing the commit, you can do this with ``git commit --amend``, which will open the commit message in an editor and let you modify it."

**Review by EZoni — COMMENTED** [review]
Thank you for this PR, Axel! I left only a few minor comments.

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
+^^^^^^^^^^^^^
 
-   .. code-block:: sh
+You are all set!
+Now, he basic WarpX development workflow is:
 
-      git checkout development
-      git pull upstream development
+1. Implement your changes and push them on a new branch ``<branch_name>`` on your fork.
+2. Create a Pull Request from branch ``<branch_name>`` on your fork to branch ``development`` on the main WarpX repo.
```
Comment (ax3l) [inline CONTRIBUTING.rst]:
```suggestion
1. Implement your changes and push them on a new branch ``branch_name`` on your fork.
2. Create a Pull Request from branch ``branch_name`` on your fork to branch ``development`` on the main WarpX repo.
```

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
 
-Make sure you are on WarpX ``development`` branch with
+Create a branch ``<branch_name>`` (the branch name should reflect the piece of code you want to add, like ``fix-spectral-solver``) with
```
Comment (ax3l) [inline CONTRIBUTING.rst]:
```suggestion
Create a branch ``branch_name`` (the branch name should reflect the piece of code you want to add, like ``fix-spectral-solver``) with
```

File: CONTRIBUTING.rst (hunk 4 above)
Comment (ax3l) [reply CONTRIBUTING.rst]: Good point. Although `<>` for mandatory and `[]` for optional arguments is standard unix convention to mark "replace this", many of our colleagues don't know that at this point and we would have to introduce it above to make sure people don't get stuck.

```suggestion
   git checkout -b branch_name
```

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
+   The most important part is really that you find a descriptive title
+   and add an empty newline after it.
 
 For the moment, commits are on your local repo only.
 You can push them to your fork with
 
 .. code-block:: sh
 
-   git push -u origin <branch_name>
+   git push -u <myGithubUsername> <branch_name>
```
Comment (ax3l) [inline CONTRIBUTING.rst]:
```suggestion
   git push -u myGithubUsername branch_name
```

File: CONTRIBUTING.rst [hunk 10]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
 
 If you want to synchronize your branch with the ``development`` branch (this is useful when the ``development`` branch is being modified while you are working on ``<branch_name>``), you can use
```
Comment (ax3l) [inline CONTRIBUTING.rst]:
```suggestion
If you want to synchronize your branch with the ``development`` branch (this is useful when the ``development`` branch is being modified while you are working on ``branch_name``), you can use
```

File: CONTRIBUTING.rst [hunk 11]
```diff
@@ -15,91 +15,126 @@ If you are new to git, you can follow one of these tutorials:
...
-   cd warpx
-   # Keep track of the main WarpX repo, to remain up-to-date.
-   git remote add upstream https://github.com/ECP-WarpX/WarpX.git
+   # rename what we just cloned: call it "mainline"
+   git remote rename origin mainline
+
+   # Add your own fork. You can get this address on your fork's Github page.
+   # Here is where you will publish new developments, so that they can be
+   # reviewed and integrated into "mainline" later on.
+   git remote add <myGithubUsername> git@github.com:<myGithubUsername>/WarpX.git
```
Comment (ax3l) [inline CONTRIBUTING.rst]:
```suggestion
   # "myGithubUsername" needs to be replaced with your user name on GitHub.
   git remote add myGithubUsername git@github.com:myGithubUsername/WarpX.git
```

File: CONTRIBUTING.rst (hunk 6 above)
Comment (ax3l) [reply CONTRIBUTING.rst:126]: I think that's too advanced for here. Multiple commits do not hurt for beginners as we squash them on merge.

History rewriting (amend, rebase) are advanced `git` topics and I don't want people to footgun themselves by starting right away with `push -f` when they are first time using `git`.

`git rm` of some files is non-destructive during review, but `push -f` can potentially remove some work that they did not backup.

**Review by EZoni — APPROVED** [review]
LGTM, thank you for updating our guidelines!

**Comment by EZoni** [comment]
I see that the Python test `Python_particle_attr_access_unique` is failing here, which is odd. Needs to be understood if the test was broken somewhere else and we just need to rebase on `development` or if it's just some kind of glitch.

**Comment by ax3l** [comment]
Likely pre-existing problem with the test, since this PR does not change the test.

Restarted CI now.

**Comment by ax3l** [comment]
Might be addressed with #2580

**Pull request merged**